docs/moderate.md

```diff
@@ -0,0 +1,193 @@
+# Moderate
+
+Let's say you want to moderate comments using OpenAI. You can do it in two ways:
+
+1. Using a trigger that will moderate the comment before it's inserted or updated in the database.
+2. Using a background action that will moderate the comments every [N configurable] seconds.
+
+## Via trigger
+
+You can get the full example in the [trigger_moderate.sql](examples/trigger_moderate.sql) file.
+
+First, let's create the extension and set the API key:
+
+```sql
+create extension if not exists ai cascade;
+```
+
+To set the API key, you can use the following command:
+```sql
+select set_config('ai.openai_api_key', :'OPENAI_API_KEY', false) is not null as set_config;
+```
+
+Or through PGOPTIONS in the command line:
+
+```bash
+PGOPTIONS="-c ai.openai_api_key=$OPENAI_API_KEY" psql -d "postgres://<username>:<password>@<host>:<port>/<database-name>"
+```
+
+So, let's create table to store the comments:
+
+```sql
+CREATE TABLE comments (
+    id SERIAL PRIMARY KEY,
+    body TEXT NOT NULL,
+    created_at TIMESTAMP NOT NULL DEFAULT NOW(),
+    status TEXT NOT NULL DEFAULT 'pending'
+);
+```
+
+Now, let's create a function that classifies the result of the openai API.
+
+```sql
+CREATE OR REPLACE FUNCTION get_moderation_status(result jsonb)
+RETURNS TEXT AS $$
+BEGIN
+    IF result->>'flagged' IS NOT NULL THEN
+        IF result->'categories'->>'violence' then
+            return 'violence';
+        END IF;
+        IF result->'categories'->>'harassment' then
+            return 'harassment';
+        END IF;
+        IF result->'categories'->>'hate' then
+            return 'hate';
+        END IF;
+        IF result->'categories'->>'sexual' then
+            return 'sexual';
+        end if;
+    end if;
+    return 'approved';
+end;
+$$ language plpgsql;
+```
+
+Creating the trigger function that changes the status of the comment based on the result of the openai API.
+
+```sql
+CREATE OR REPLACE FUNCTION moderate_comment() RETURNS TRIGGER AS $$
+declare
+    out jsonb;
+BEGIN
+  select openai_moderate(
+    'text-moderation-stable',
+    NEW.body
+  )->'results'->0 into out;
+  NEW.status = get_moderation_status(out);
+
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
```

Comment:
This trigger only works if the `ai.openai_api_key` session parameter is set. If not, the insert will fail and roll back. Maybe that's a good thing? We should probably note that. It might be worth setting the API key explicitly to make it more obvious, like:

```sql
  select openai_moderate(
    'text-moderation-stable',
    NEW.body,
    _api_key=>current_setting('ai.openai_api_key', false) -- fail if setting not available
  )->'results'->0 into out;
  NEW.status = get_moderation_status(out);
```

Reply:
Added.